zipline-kotlin-plugin/src/main/kotlin/app/cash/zipline/ktbridge/plugin/BridgedInterface.kt

```diff
@@ -17,28 +17,112 @@ package app.cash.zipline.ktbridge.plugin
 import org.jetbrains.kotlin.ir.util.substitute
 import org.jetbrains.kotlin.name.FqName
+import org.jetbrains.kotlin.name.Name
 
 /**
  * A user-defined interface (like `EchoService` or `Callback<String>`) and support for either
  * implementing it ([KtBridgeGetRewriter]) or calling it ([KtBridgeSetRewriter]).
  *
  * This class tracks the interface type (like `EchoService` or `Callback<String>`) and its
  * implementation class (that doesn't know its generic parameters).
+ *
+ * This class can declare a `KSerializer` property for each unique parameter type and return type on
+ * the interface. Use [declareSerializerProperties] to create these, then [serializerExpression] to
+ * access a specific serializer. Declaring serializers is useful to fast fail if ever a serializer
+ * is required but not configured.
  */
 internal class BridgedInterface(
+  private val pluginContext: IrPluginContext,
+  private val ktBridgeApis: KtBridgeApis,
+
   /** A specific type identifier that knows the values of its generic parameters. */
   val type: IrType,
 
   /** A potentially-generic declaration that doesn't have values for its generic parameters. */
-  val classSymbol: IrClassSymbol,
+  private val classSymbol: IrClassSymbol,
 ) {
+  /** Maps types to the property holding the corresponding serializer. */
+  private val typeToSerializerProperty = mutableMapOf<IrType, IrProperty>()
+
+  val bridgedFunctions: List<IrSimpleFunctionSymbol>
+    // TODO(jwilson): find a better way to skip equals()/hashCode()/toString()
+    get() = classSymbol.functions.toList()
+      .filterNot { it.owner.isFakeOverride }
+
+  /** Declares properties for all the serializers needed to bridge this interface. */
+  fun declareSerializerProperties(declaringClass: IrClass) {
+    check(typeToSerializerProperty.isEmpty()) { "declareSerializerProperties() called twice?" }
+
+    val serializedTypes = mutableSetOf<IrType>()
+    for (bridgedFunction in bridgedFunctions) {
+      for (valueParameter in bridgedFunction.owner.valueParameters) {
+        serializedTypes += resolveTypeParameters(valueParameter.type)
+      }
+      serializedTypes += resolveTypeParameters(bridgedFunction.owner.returnType)
+    }
+
+    for (serializedType in serializedTypes) {
+      val serializerProperty = irSerializerProperty(
+        declaringClass = declaringClass,
+        type = serializedType,
+        name = Name.identifier("serializer_${typeToSerializerProperty.size}")
+      )
+      declaringClass.declarations += serializerProperty
+      typeToSerializerProperty[serializedType] = serializerProperty
+    }
+  }
+
+  private fun irSerializerProperty(
+    declaringClass: IrClass,
+    type: IrType,
+    name: Name
+  ): IrProperty {
+    val serializersModuleProperty = declaringClass.properties.single {
+      it.getter?.returnType?.classFqName == ktBridgeApis.serializersModuleFqName
+    }
+
+    // val serializer_0: KSerializer<EchoRequest> = serializersModule.serializer<EchoRequest>()
```

Comment:
these docs are fantastic and make the code much more readable to see what's happening